File: week5/c++/1_money_change_again.cpp
#include <iostream>
#include <vector>

using namespace std;

int DPchange(int money, vector<int> coins)
{
    vector<int> minnumcoins(money + 1);
    minnumcoins[0] = 0;
    for (int m = 1; m <= money; m++)
    {
        minnumcoins[m] = 10000;
        for (auto c : coins)
        {
            if (m >= c)
            {
                /*int numcoins = minnumcoins[m - c] + 1;
                if (numcoins < minnumcoins[m])
                {
                    minnumcoins[m] = numcoins;
                }
                */
                minnumcoins[m] = min(minnumcoins[m], minnumcoins[m - c] + 1);
            }
        }
    }
    return minnumcoins[money];
}

int main()
{
    int n;
    cin >> n;

    vector<int> v{1, 3, 4};
    cout << DPchange(n, v) << endl;
    return 0;
}

File: week5/python/1_money_change_again.py
'''
def DPchange(money, coins):
    minnumcoins = dict()
    minnumcoins[0] = 0
    for m in range(1, money+1):
        minnumcoins[m] = 10**4
        for i in range(len(coins)):
            if m >= coins[i]:
                numcoins = minnumcoins[m-coins[i]] + 1
                if numcoins < minnumcoins[m]:
                    minnumcoins[m] = numcoins
    return minnumcoins[money]
'''


def DPchange(money, coins):
    minnumcoins = dict()
    minnumcoins[0] = 0
    for m in range(1, money+1):
        minnumcoins[m] = 10**4
        for c in coins:
            if m >= c:
                minnumcoins[m] = min(minnumcoins[m], minnumcoins[m-c]+1)
    return minnumcoins[money]


print(DPchange(int(input()), [1, 3, 4]))
